presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationPushDown.java

```diff
@@ -0,0 +1,205 @@
+                .setSystemProperty("kafka.timestamp_upper_bound_force_push_down_enabled", "true")
+                .build();
+
+        queryResult = queryRunner.executeWithQueryId(enableTmUpperBoundPushDownSession, sql);
+        assertEquals(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions(), 4);
+    }
+
+    @Test
+    public void testTimestampLogAppendModePushDown()
+    {
+        Pair<Long, Long> timePair = createTimestampTestMessages(topicNameLogAppend);
+        DistributedQueryRunner queryRunner = (DistributedQueryRunner) getQueryRunner();
+        // ">= from_unixtime(%s)" insure including index 2, "< from_unixtime(%s)"  insure excluding index 4;
+        String sql = String.format("SELECT count(*) FROM default.%s WHERE _timestamp >= from_unixtime(%s) and _timestamp < from_unixtime(%s)",
+                topicNameLogAppend, timePair.first() / 1000, timePair.second() / 1000);
+        ResultWithQueryId<MaterializedResult> queryResult = queryRunner.executeWithQueryId(getSession(), sql);
+        assertEquals(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions(), 4);
+    }
+
+    private QueryInfo getQueryInfo(DistributedQueryRunner queryRunner, ResultWithQueryId<MaterializedResult> queryResult)
+    {
+        return queryRunner.getCoordinator().getQueryManager().getFullQueryInfo(queryResult.getQueryId());
+    }
+
+    private Pair<Long, Long> createTimestampTestMessages(String topicName)
+    {
+        long startTime = -1;
+        long endTime = -1;
+        try (KafkaProducer<Long, Object> producer = testingKafka.createProducer()) {
+            for (long j = 0; j < MESSAGE_NUM; j++) {
+                Future<RecordMetadata> send = producer.send(new ProducerRecord<>(topicName, j, j));
+                try {
+                    if (j < TIMESTAMP_TEST_COUNT) {
+                        if (j == TIMESTAMP_TEST_START_INDEX) {
+                            RecordMetadata r = send.get();
+                            startTime = r.timestamp();
+                        }
+                        else if (j == TIMESTAMP_TEST_END_INDEX) {
+                            RecordMetadata r = send.get();
+                            endTime = r.timestamp();
+                        }
+                        else {
+                            send.get();
+                        }
+                        Thread.sleep(1100);
+                    }
+                }
+                catch (Exception e) {
+                    e.printStackTrace();
+                }
+            }
+        }
+        return Pair.of(startTime, endTime);
+    }
+
+    private void createMessages(String topicName)
+    {
+        try (KafkaProducer<Long, Object> producer = testingKafka.createProducer()) {
+            for (long j = 0; j < MESSAGE_NUM; j++) {
+                Future<RecordMetadata> send = producer.send(new ProducerRecord<>(topicName, j, j));
+            }
+        }
+    }
```

Comment:
```suggestion
    private void createMessages(String topicName)
            throws ExecutionException, InterruptedException
    {
        Future<RecordMetadata> lastSendFuture = Futures.immediateFuture(null);
        try (KafkaProducer<Long, Object> producer = testingKafka.createProducer()) {
            for (long messageNum = 0; messageNum < MESSAGE_NUM; messageNum++) {
                long key = messageNum;
                long value = messageNum;
                lastSendFuture = producer.send(new ProducerRecord<>(topicName, key, value));
            }
        }
        lastSendFuture.get();
    }
```

A little more code makes it more readable IMO